trunk/user/samba3/docs/manpages/umount.cifs.8
.\"     Title: umount.cifs
.\"    Author: [see the "AUTHOR" section]
.\" Generator: DocBook XSL Stylesheets v1.74.0 <http://docbook.sf.net/>
.\"      Date: 09/30/2009
.\"    Manual: System Administration tools
.\"    Source: Samba 3.0
.\"  Language: English
.\"
.TH "UMOUNT\&.CIFS" "8" "09/30/2009" "Samba 3\&.0" "System Administration tools"
.\" -----------------------------------------------------------------
.\" * (re)Define some macros
.\" -----------------------------------------------------------------
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" toupper - uppercase a string (locale-aware)
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.de toupper
.tr aAbBcCdDeEfFgGhHiIjJkKlLmMnNoOpPqQrRsStTuUvVwWxXyYzZ
\\$*
.tr aabbccddeeffgghhiijjkkllmmnnooppqqrrssttuuvvwwxxyyzz
..
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" SH-xref - format a cross-reference to an SH section
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.de SH-xref
.ie n \{\
.\}
.toupper \\$*
.el \{\
\\$*
.\}
..
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" SH - level-one heading that works better for non-TTY output
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.de1 SH
.\" put an extra blank line of space above the head in non-TTY output
.if t \{\
.sp 1
.\}
.sp \\n[PD]u
.nr an-level 1
.set-an-margin
.nr an-prevailing-indent \\n[IN]
.fi
.in \\n[an-margin]u
.ti 0
.HTML-TAG ".NH \\n[an-level]"
.it 1 an-trap
.nr an-no-space-flag 1
.nr an-break-flag 1
\." make the size of the head bigger
.ps +3
.ft B
.ne (2v + 1u)
.ie n \{\
.\" if n (TTY output), use uppercase
.toupper \\$*
.\}
.el \{\
.nr an-break-flag 0
.\" if not n (not TTY), use normal case (not uppercase)
\\$1
.in \\n[an-margin]u
.ti 0
.\" if not n (not TTY), put a border/line under subheading
.sp -.6
\l'\n(.lu'
.\}
..
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" SS - level-two heading that works better for non-TTY output
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.de1 SS
.sp \\n[PD]u
.nr an-level 1
.set-an-margin
.nr an-prevailing-indent \\n[IN]
.fi
.in \\n[IN]u
.ti \\n[SN]u
.it 1 an-trap
.nr an-no-space-flag 1
.nr an-break-flag 1
.ps \\n[PS-SS]u
\." make the size of the head bigger
.ps +2
.ft B
.ne (2v + 1u)
.if \\n[.$] \&\\$*
..
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" BB/BE - put background/screen (filled box) around block of text
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.de BB
.if t \{\
.sp -.5
.br
.in +2n
.ll -2n
.gcolor red
.di BX
.\}
..
.de EB
.if t \{\
.if "\\$2"adjust-for-leading-newline" \{\
.sp -1
.\}
.br
.di
.in
.ll
.gcolor
.nr BW \\n(.lu-\\n(.i
.nr BH \\n(dn+.5v
.ne \\n(BHu+.5v
.ie "\\$2"adjust-for-leading-newline" \{\
\M[\\$1]\h'1n'\v'+.5v'\D'P \\n(BWu 0 0 \\n(BHu -\\n(BWu 0 0 -\\n(BHu'\M[]
.\}
.el \{\
\M[\\$1]\h'1n'\v'-.5v'\D'P \\n(BWu 0 0 \\n(BHu -\\n(BWu 0 0 -\\n(BHu'\M[]
.\}
.in 0
.sp -.5v
.nf
.BX
.in
.sp .5v
.fi
.\}
..
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" BM/EM - put colored marker in margin next to block of text
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.de BM
.if t \{\
.br
.ll -2n
.gcolor red
.di BX
.\}
..
.de EM
.if t \{\
.br
.di
.ll
.gcolor
.nr BH \\n(dn
.ne \\n(BHu
\M[\\$1]\D'P -.75n 0 0 \\n(BHu -(\\n[.i]u - \\n(INu - .75n) 0 0 -\\n(BHu'\M[]
.in 0
.nf
.BX
.in
.fi
.\}
..
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "Name"
umount.cifs \- for normal, non\-root users, to unmount their own Common Internet File System (CIFS) mounts
.SH "Synopsis"
.fam C
.HP \w'\ 'u
\FCumount\&.cifs\F[] {mount\-point} [\-nVvhfle]
.fam
.SH "DESCRIPTION"
.PP
This tool is part of the
\fBsamba\fR(7)
suite\&.
.PP
umount\&.cifs unmounts a Linux CIFS filesystem\&. It can be invoked indirectly by the
\fBumount\fR(8)
command when umount\&.cifs is in /sbin directory, unless you specify the "\-i" option to umount\&. Specifying \-i to umount avoids execution of umount helpers such as umount\&.cifs\&. The umount\&.cifs command only works in Linux, and the kernel must support the cifs filesystem\&. The CIFS protocol is the successor to the SMB protocol and is supported by most Windows servers and many other commercial servers and Network Attached Storage appliances as well as by the popular Open Source server Samba\&.
.PP
The umount\&.cifs utility detaches the local directory
\fImount\-point\fR
from the corresponding UNC name (exported network resource) and frees the associated kernel resources\&. It is possible to set the mode for umount\&.cifs to setuid root (or equivalently update the /etc/permissions file) to allow non\-root users to umount shares to directories for which they have write permission\&. The umount\&.cifs utility is typically not needed if unmounts need only be performed by root users, or if user mounts and unmounts can rely on specifying explicit entries in /etc/fstab See
.PP
\fBfstab\fR(5)
.SH "OPTIONS"
.PP
\-\-verbose
.RS 4
print additional debugging information
.RE
.PP
\-\-no\-mtab
.RS 4
Do not update the mtab even if unmount completes successfully (/proc/mounts will still display the correct information)
.RE
.SH "NOTES"
.PP
This command is normally intended to be installed setuid (since root users can already run unmount)\&. An alternative to using umount\&.cifs is to add specfic entries for the user mounts that you wish a particular user or users to mount and unmount to /etc/fstab
.SH "CONFIGURATION"
.PP
The primary mechanism for making configuration changes and for reading debug information for the cifs vfs is via the Linux /proc filesystem\&. In the directory
\FC/proc/fs/cifs\F[]
are various configuration files and pseudo files which can display debug information\&. For more information see the kernel file
\FCfs/cifs/README\F[]\&.
.SH "BUGS"
.PP
At this time umount\&.cifs does not lock the mount table using the same lock as the umount utility does, so do not attempt to do multiple unmounts from different processes (and in particular unmounts of a cifs mount and another type of filesystem mount at the same time)\&.
.PP
If the same mount point is mounted multiple times by cifs, umount\&.cifs will remove all of the matching entries from the mount table (although umount\&.cifs will actually only unmount the last one), rather than only removing the last matching entry in /etc/mtab\&. The pseudofile /proc/mounts will display correct information though, and the lack of an entry in /etc/mtab does not prevent subsequent unmounts\&.
.PP
Note that the typical response to a bug report is a suggestion to try the latest version first\&. So please try doing that first, and always include which versions you use of relevant software when reporting bugs (minimum: umount\&.cifs (try umount\&.cifs \-V), kernel (see /proc/version) and server type you are trying to contact\&.
.SH "VERSION"
.PP
This man page is correct for version 1\&.34 of the cifs vfs filesystem (roughly Linux kernel 2\&.6\&.12)\&.
.SH "SEE ALSO"
.PP
Documentation/filesystems/cifs\&.txt and fs/cifs/README in the linux kernel source tree may contain additional options and information\&.
.PP
\fBmount.cifs\fR(8)
.SH "AUTHOR"
.PP
Steve French
.PP
The syntax was loosely based on the umount utility and the manpage was loosely based on that of mount\&.cifs\&.8\&. The man page was created by Steve French
.PP
The maintainer of the Linux cifs vfs and the userspace tool
\fIumount\&.cifs\fR
is
Steve French\&. The
Linux CIFS Mailing list
is the preferred place to ask questions regarding these programs\&.


trunk/user/samba3/source/client/umount.cifs.c
/* 
   Unmount utility program for Linux CIFS VFS (virtual filesystem) client
   Copyright (C) 2005 Steve French  (sfrench@us.ibm.com)

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2 of the License, or
   (at your option) any later version.
   
   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.
   
   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software
   Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.  */

#ifndef _GNU_SOURCE
#define _GNU_SOURCE
#endif

#include <stdlib.h>
#include <stdio.h>
#include <unistd.h>
#include <ctype.h>
#include <sys/types.h>
#include <sys/mount.h>
#include <sys/ioctl.h>
#include <sys/stat.h>
#include <sys/vfs.h>
#include <fcntl.h>
#include <getopt.h>
#include <errno.h>
#include <string.h>
#include <mntent.h>
#include <limits.h>
#include "mount.h"

#define UNMOUNT_CIFS_VERSION_MAJOR "0"
#define UNMOUNT_CIFS_VERSION_MINOR "6"

#ifndef UNMOUNT_CIFS_VENDOR_SUFFIX
 #ifdef _SAMBA_BUILD_
  #include "include/version.h"
  #ifdef SAMBA_VERSION_VENDOR_SUFFIX
   #define UNMOUNT_CIFS_VENDOR_SUFFIX "-"SAMBA_VERSION_OFFICIAL_STRING"-"SAMBA_VERSION_VENDOR_SUFFIX
  #else
   #define UNMOUNT_CIFS_VENDOR_SUFFIX "-"SAMBA_VERSION_OFFICIAL_STRING
  #endif /* SAMBA_VERSION_OFFICIAL_STRING and SAMBA_VERSION_VENDOR_SUFFIX */
 #else
  #define UNMOUNT_CIFS_VENDOR_SUFFIX ""
 #endif /* _SAMBA_BUILD_ */
#endif /* UNMOUNT_CIFS_VENDOR_SUFFIX */

#ifndef MNT_DETACH
#define MNT_DETACH 0x02
#endif

#ifndef MNT_EXPIRE
#define MNT_EXPIRE 0x04
#endif

#ifndef MOUNTED_LOCK
#define MOUNTED_LOCK    "/etc/mtab~"
#endif
#ifndef MOUNTED_TEMP
#define MOUNTED_TEMP    "/etc/mtab.tmp"
#endif

#define CIFS_IOC_CHECKUMOUNT _IO(0xCF, 2)
#define CIFS_MAGIC_NUMBER 0xFF534D42   /* the first four bytes of SMB PDU */
   
static struct option longopts[] = {
	{ "all", 0, NULL, 'a' },
	{ "help",0, NULL, 'h' },
	{ "read-only", 0, NULL, 'r' },
	{ "ro", 0, NULL, 'r' },
	{ "verbose", 0, NULL, 'v' },
	{ "version", 0, NULL, 'V' },
	{ "expire", 0, NULL, 'e' },
	{ "force", 0, 0, 'f' },
	{ "lazy", 0, 0, 'l' },
	{ "no-mtab", 0, 0, 'n' },
	{ NULL, 0, NULL, 0 }
};

const char * thisprogram;
int verboseflg = 0;

static void umount_cifs_usage(void)
{
	printf("\nUsage:  %s <remotetarget> <dir>\n", thisprogram);
	printf("\nUnmount the specified directory\n");
	printf("\nLess commonly used options:");
	printf("\n\t-r\tIf mount fails, retry with readonly remount.");
	printf("\n\t-n\tDo not write to mtab.");
	printf("\n\t-f\tAttempt a forced unmount, even if the fs is busy.");
	printf("\n\t-l\tAttempt lazy unmount, Unmount now, cleanup later.");
	printf("\n\t-v\tEnable verbose mode (may be useful for debugging).");
	printf("\n\t-h\tDisplay this help.");
	printf("\n\nOptions are described in more detail in the manual page");
	printf("\n\tman 8 umount.cifs\n");
	printf("\nTo display the version number of the cifs umount utility:");
	printf("\n\t%s -V\n",thisprogram);
	printf("\nInvoking the umount utility on cifs mounts, can execute");
	printf(" /sbin/umount.cifs (if present and umount -i is not specified.\n");
}

static int umount_check_perm(char * dir)
{
	int fileid;
	int rc;

	/* allow root to unmount, no matter what */
	if(getuid() == 0)
		return 0;

	/* presumably can not chdir into the target as we do on mount */
	fileid = open(dir, O_RDONLY | O_DIRECTORY | O_NOFOLLOW, 0);
	if(fileid == -1) {
		if(verboseflg)
			printf("error opening mountpoint %d %s",errno,strerror(errno));
		return errno;
	}

	rc = ioctl(fileid, CIFS_IOC_CHECKUMOUNT, NULL);

	if(verboseflg)
		printf("ioctl returned %d with errno %d %s\n",rc,errno,strerror(errno));

	if(rc == ENOTTY) {
		printf("user unmounting via %s is an optional feature of",thisprogram);
		printf(" the cifs filesystem driver (cifs.ko)");
		printf("\n\tand requires cifs.ko version 1.32 or later\n");
	} else if (rc != 0)
		printf("user unmount of %s failed with %d %s\n",dir,errno,strerror(errno));
	close(fileid);

	return rc;
}

static int remove_from_mtab(char * mountpoint)
{
	int rc;
	int num_matches;
	FILE * org_fd;
	FILE * new_fd;
	struct mntent * mount_entry;

	/* Do we need to check if it is a symlink to e.g. /proc/mounts
	in which case we probably do not want to update it? */

	/* Do we first need to check if it is writable? */ 

	atexit(unlock_mtab);
	if (lock_mtab()) {
		printf("Mount table locked\n");
		return -EACCES;
	}
	
	if(verboseflg)
		printf("attempting to remove from mtab\n");

	org_fd = setmntent(MOUNTED, "r");

	if(org_fd == NULL) {
		printf("Can not open %s\n",MOUNTED);
		unlock_mtab();
		return -EIO;
	}

	new_fd = setmntent(MOUNTED_TEMP,"w");
	if(new_fd == NULL) {
		printf("Can not open temp file %s", MOUNTED_TEMP);
		endmntent(org_fd);
		unlock_mtab();
		return -EIO;
	}

	/* BB fix so we only remove the last entry that matches BB */
	num_matches = 0;
	while((mount_entry = getmntent(org_fd)) != NULL) {
		if(strcmp(mount_entry->mnt_dir, mountpoint) == 0) {
			num_matches++;
		}
	}	
	if(verboseflg)
		printf("%d matching entries in mount table\n", num_matches);
		
	/* Is there a better way to seek back to the first entry in mtab? */
	endmntent(org_fd);
	org_fd = setmntent(MOUNTED, "r");

	if(org_fd == NULL) {
		printf("Can not open %s\n",MOUNTED);
		unlock_mtab();
		return -EIO;
	}
	
	while((mount_entry = getmntent(org_fd)) != NULL) {
		if(strcmp(mount_entry->mnt_dir, mountpoint) != 0) {
			addmntent(new_fd, mount_entry);
		} else {
			if(num_matches != 1) {
				addmntent(new_fd, mount_entry);
				num_matches--;
			} else if(verboseflg)
				printf("entry not copied (ie entry is removed)\n");
		}
	}

	if(verboseflg)
		printf("done updating tmp file\n");
	rc = fchmod (fileno (new_fd), S_IRUSR|S_IWUSR|S_IRGRP|S_IROTH);
	if(rc < 0) {
		printf("error %s changing mode of %s\n", strerror(errno),
			MOUNTED_TEMP);
	}
	endmntent(new_fd);

	rc = rename(MOUNTED_TEMP, MOUNTED);

	if(rc < 0) {
		printf("failure %s renaming %s to %s\n",strerror(errno),
			MOUNTED_TEMP, MOUNTED);
		unlock_mtab();
		return -EIO;
	}

	unlock_mtab();
	
	return rc;
}

/* Make a canonical pathname from PATH.  Returns a freshly malloced string.
   It is up the *caller* to ensure that the PATH is sensible.  i.e.
   canonicalize ("/dev/fd0/.") returns "/dev/fd0" even though ``/dev/fd0/.''
   is not a legal pathname for ``/dev/fd0''  Anything we cannot parse
   we return unmodified.   */
static char *
canonicalize(char *path)
{
	char *canonical = malloc (PATH_MAX + 1);

	if (!canonical) {
		fprintf(stderr, "Error! Not enough memory!\n");
		return NULL;
	}

	if (strlen(path) > PATH_MAX) {
		fprintf(stderr, "Mount point string too long\n");
		return NULL;
	}

	if (path == NULL)
		return NULL;

	if (realpath (path, canonical))
		return canonical;

	strncpy (canonical, path, PATH_MAX);
	canonical[PATH_MAX] = '\0';
	return canonical;
}

int main(int argc, char ** argv)
{
	int c;
	int rc;
	int flags = 0;
	int nomtab = 0;
	int retry_remount = 0;
	struct statfs statbuf;
	char * mountpoint;

	if(argc && argv) {
		thisprogram = argv[0];
	} else {
		umount_cifs_usage();
		return -EINVAL;
	}

	if(argc < 2) {
		umount_cifs_usage();
		return -EINVAL;
	}

	if(thisprogram == NULL)
		thisprogram = "umount.cifs";

	/* add sharename in opts string as unc= parm */

	while ((c = getopt_long (argc, argv, "afhilnrvV",
			 longopts, NULL)) != -1) {
		switch (c) {
/* No code to do the following  option yet */
/*		case 'a':	       
			++umount_all;
			break; */
		case '?':
		case 'h':   /* help */
			umount_cifs_usage();
			exit(1);
		case 'n':
			++nomtab;
			break;
		case 'f':
			flags |= MNT_FORCE;
			break;
		case 'l':
			flags |= MNT_DETACH; /* lazy unmount */
			break;
		case 'e':
			flags |= MNT_EXPIRE; /* gradually timeout */
			break;
		case 'r':
			++retry_remount;
			break;
		case 'v':
			++verboseflg;
			break;
		case 'V':	   
			printf ("umount.cifs version: %s.%s%s\n",
				UNMOUNT_CIFS_VERSION_MAJOR,
				UNMOUNT_CIFS_VERSION_MINOR,
				UNMOUNT_CIFS_VENDOR_SUFFIX);
			exit (0);
		default:
			printf("unknown unmount option %c\n",c);
			umount_cifs_usage();
			exit(1);
		}
	}

	/* move past the umount options */
	argv += optind;
	argc -= optind;

	mountpoint = canonicalize(argv[0]);

	if((argc < 1) || (argv[0] == NULL)) {
		printf("\nMissing name of unmount directory\n");
		umount_cifs_usage();
		return -EINVAL;
	}

	if(verboseflg)
		printf("optind %d unmount dir %s\n",optind, mountpoint);

	/* check if running effectively root */
	if(geteuid() != 0) {
		printf("Trying to unmount when %s not installed suid\n",thisprogram);
		if(verboseflg)
			printf("euid = %d\n",geteuid());
		return -EACCES;
	}

	/* fixup path if needed */

	/* Trim any trailing slashes */
	while ((strlen(mountpoint) > 1) &&
		(mountpoint[strlen(mountpoint)-1] == '/'))
	{
		mountpoint[strlen(mountpoint)-1] = '\0';
	}

	/* make sure that this is a cifs filesystem */
	rc = statfs(mountpoint, &statbuf);
	
	if(rc || (statbuf.f_type != CIFS_MAGIC_NUMBER)) {
		printf("This utility only unmounts cifs filesystems.\n");
		return -EINVAL;
	}

	/* check if our uid was the one who mounted */
	rc = umount_check_perm(mountpoint);
	if (rc) {
		printf("Not permitted to unmount\n");
		return rc;
	}

	if(umount2(mountpoint, flags)) {
	/* remember to kill daemon on error */
		switch (errno) {
		case 0:
			printf("unmount failed but no error number set\n");
			break;
		default:
			printf("unmount error %d = %s\n",errno,strerror(errno));
		}
		printf("Refer to the umount.cifs(8) manual page (man 8 umount.cifs)\n");
		return -1;
	} else {
		if(verboseflg)
			printf("umount2 succeeded\n");
		if(nomtab == 0)
			remove_from_mtab(mountpoint);
	}

	return 0;
}

